### PR TITLE
Add nRF Cloud CoAP download support

### PR DESCRIPTION
**Pull request by jayteemo**
Update `fota_download` to allow for FOTA updates to be performed using an external download source.
Update `nrf_cloud` library to enable downloading FOTA and P-GPS data using nRF Cloud's CoAP proxy download resource.

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_download.c [hunk 1]
```diff
@@ -9,17 +9,271 @@
...
+		return ret;
+	}
+
+	ret = coap_dl_connect_and_auth(dl);
+	if (ret) {
+		/* ERROR TODO */
+		return ret;
+	}
+
+	/* Get the options for the proxy download */
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_download.c:205]: TODO: 
Need to add a call to fota_download_b1_file_parse before getting the options.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_download.c [hunk 2]
```diff
@@ -9,20 +9,300 @@
...
+
+	return -EINVAL;
+}
+
+static int coap_dl_disconnect(void)
+{
+	return nrf_cloud_coap_transport_disconnect(&coap_client);
+}
+
+static void coap_cb(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_download.c:89]: Maybe coap_dl_cb?

File: include/net/fota_download.h [hunk 3]
```diff
@@ -260,6 +267,46 @@ int fota_download_target(void);
...
+
+/**@brief Handle a download event from an external download client.
+ *        Requires @kconfig{CONFIG_FOTA_DOWNLOAD_EXTERNAL_DL} to be enabled.
+ *
+ * @param evt Download event.
+ *
+ * @retval 0 If successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int fota_download_external_evt_handle(struct download_client_evt const *const evt);
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply include/net/fota_download.h:308]: yes i thought about making a new event struct... in the end it didn't seem worth it.
the download client events need to be duplicated, so there would just be a straight translation in `fota_download_external_evt_handle`.
let's see what the code owner think.

File: subsys/net/lib/fota_download/src/fota_download.c [hunk 4]
```diff
@@ -366,6 +423,44 @@ static bool is_ip_address(const char *host)
...
+	const char *update;
+	bool s0_active;
+	int err;
+
+	err = fota_download_s0_active_get(&s0_active);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	err = fota_download_parse_dual_resource_locator(s0_s1_files, s0_active, &update);
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/fota_download/src/fota_download.c:444]: i'd rather not remove the existing API `fota_download_parse_dual_resource_locator` (if that is what you mean).
again i will defer to the code owners opinion.

**Review by plskeggs — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
I'm lost in the code...

What exactly is the aim here? This seem to refactor the fota_download client so much.

The current design was quite simple
```
                                                                           
                                                                           
                                                                           
                              ┌──────────┐                                 
                              │DFU_TARGET│                                 
                              └──────────┘                                 
                                   ▲                                       
                                   │                                       
         ┌────┐  start      ┌──────┴────────┐ start   ┌────────────────┐   
         │APP ├────────────►│ fota_download ├────────►│download_client │   
         └────┘             └───┬───────────┘         └────────┬───────┘   
             ▲                  │        ▲                     │           
             └──────────────────┘        │                     │           
               fota_download_evt         └─────────────────────┘           
                                            download_client_evt            
                                                                           
                                                                           
```

Pretty much just a wrapper that uses download_client and push the data into DFU_target library.

But after this refactoring I'm not sure what exactly is the role of fota_download? This "external downloader" confuses me. If the nRF Cloud library downloads the FOTA pieces, why don't you directly push those into dfu_target? Why the data has to go through fota_download library?

```
                                                                          
                      ┌──────────┐                                        
                      │DFU_TARGET│                                        
                      └──────────┘                                        
                           ▲                                              
                           │                                              
 ┌────┐  start      ┌──────┴────────┐ start   ┌────────────────┐          
 │APP ├────────────►│ fota_download ├────────►│download_client │          
 └────┘             └───┬───────────┘         └────────┬───────┘          
     ▲                  │ ▲    ▲ ▲                     │                  
     └──────────────────┘ │    │ │                     │                  
       fota_download_evt  │    │ └─────────────────────┘                  
                          │    │    download_client_evt                   
                          │    │                                          
                          │    │                                          
             external_start    │external_evt                              
                          │    │                                          
                          │  ┌─┴────────────┐start   ┌──────────────────┐ 
                          └──┤nrf_cloud_fota├───────►│nrf_cloud_download│ 
                             └──────────────┘        └──────┬───────────┘ 
    ┌───┐                       ▲         ▲   evt           │             
    │APP├───────────────────────┘         └─────────────────┘             
    └───┘       use                                                       
                                                                          
```

Why not just
```
                                                                           
                                                                           
                                                                           
                          ┌──────────┐                                     
                          │DFU_TARGET│                                     
                          └──────────┘                                     
                               ▲                                           
                               │                                           
                               │                                           
                             ┌─┴────────────┐start   ┌──────────────────┐  
                             │nrf_cloud_fota├───────►│nrf_cloud_download│  
                             └──────────────┘        └──────┬───────────┘  
    ┌───┐                       ▲         ▲   evt           │              
    │APP├───────────────────────┘         └─────────────────┘              
    └───┘       use                                                        
                                                                           
```

The reason why I would hesitate to refactor the fota_download library is that there has been so much undocumented assumptions on how it deal on certain events coming from downloader and what to do on dfu_target on those. Many times it has been almost trial and error to get it right. Like, when to use the `dfu_target_done(false);` instead of `dfu_target_reset()`.
Now that if fota_downloader becomes a wrapper between two libraries, I fear that we start to cumulate too much of these if-this-then-that translations between events. Then it might become unmaintainable as all the details of certain use cases are lost inside complex looking code.

**Comment by jayteemo** [comment]
> What exactly is the aim here? This seem to refactor the fota_download client so much.

We need a way to perform FOTA over CoAP. 
nRF Cloud CoAP provides a resource to perform file downloads. This cannot be performed using `download_client`.
Initially, I looked into updating `download_client` and even using `coap_client` within `download_client`, but that turned out to be an even larger change.
So we are able to use `coap_client` to perform the download, but then we would need to re-implement all of the FOTA installation, error and event handling... And as you mentioned:

> there has been so much undocumented assumptions on how it deal on certain events coming from downloader and what to do on dfu_target

We would then be duplicating all of that in the `nrf_cloud` library.

There are no functionality changes for the normal usage of `fota_download`.  I just moved code into functions.

**Review by glarsennordic — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 5]
```diff
@@ -964,6 +965,13 @@ Libraries for networking
 
   * Removed the deprecated ``download_client_connect`` function.
 
+* :ref:`lib_fota_download` library:
+
+  * Added:
+
+    * The function :c:func:`fota_download_b1_file_parse` to parse a bootloader update file path.
+    * Experimental support for performing FOTA updates using an external download client with the Kconfig option :kconfig:option:`CONFIG_FOTA_DOWNLOAD_EXTERNAL_DL` and functions :c:func:`fota_download_external_start` and Function :c:func:`fota_download_external_evt_handle`.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:973]:
```suggestion
    * The :c:func:`fota_download_b1_file_parse` function to parse a bootloader update file path.
    * Experimental support for performing FOTA updates using an external download client with the :kconfig:option:`CONFIG_FOTA_DOWNLOAD_EXTERNAL_DL` Kconfig option and functions :c:func:`fota_download_external_start` and :c:func:`fota_download_external_evt_handle`.
```

**Review by peknis — APPROVED** [review]
Approved with a couple of minor nits.

**Pull request merged**